Development/lab 7

notjoncruz opened the pull request:
- [x] No discrepancies across browsers. Eg. chrome vs safari
- [ ] Pages pass W3C Validation for HTML (https://validator.w3.org)
- [ ] CSS (https://jigsaw.w3.org/css-validator/)
- [x] Google sign-in is operational
- [x] Accessibility functions
- [x] Pages can scale without distorting page
- [x] No dead links
- [x] Navbar is consistent across the site
- [x] Pages are screen-reader accessible
- [x] Contrast meets standards for accessibility
- [x] Pages are keyboard accessible
- [ ] Passes WAVE Evaluation Tool chrome extension

kxl1360 committed:
Merge branch 'dev' of https://github.com/all-rit/ALL into development/Lab-7

kxl1360 committed:
stubbed out pages for lab 7 and added lab7 route to routes/index.js

kxl1360 committed:
rm

kxl1360 committed:
create lab7 client side

kxl1360 committed:
working url to Lab 7 page

kxl1360 committed:
more skeleton pages

kxl1360 committed:
lab 7 server files

kxl1360 committed:
Lab 7 reading in database and created exercise pages

kxl1360 committed:
create rest of exercise pages and fixed routing

kxl1360 committed:
create quiz skeleton

kxl1360 committed:
small formatting tweaks

kxl1360 committed:
create components and added them to pages

kxl1360 committed:
create missing main components and fixed reducers for lab7

kxl1360 committed:
minor css change to sim summary

kxl1360 committed:
refined components and add imgs for sim

kxl1360 committed:
create missing sim components and add scss file

kxl1360 committed:
sim ui 90% done, mostly css work

kxl1360 committed:
fix button css

kxl1360 committed:
fix button css pt 2

kxl1360 committed:
simulation update

kxl1360 committed:
found issue with file reset after each round

kxl1360 committed:
repair done

kxl1360 committed:
added quiz and changed lab7 exercise table

kxl1360 committed:
add alteration exercise quiz

kxl1360 committed:
added comment to repair and found simulation running issues found in repair part of exercise

kxl1360 committed:
fixed naming issues

kxl1360 committed:
fixed typos and parameter mistakes

kxl1360 committed:
pushing missed file

Commit:
Merged from dev and fixed conflicts

Commit:
Handling eslint errors

jjm8759 committed:
Added endpoint for Lab7 exercise end because API.js was returning 404 to client

Commit:
[#142] Updated components, styling, and round animation.

Commit:
[#142] Added functional progress bar and message for when the AI is incorrect.

Commit:
[#142] Functional simulation of all ten rounds.

Commit:
[#142] Functional tally system for all ten rounds.

dm9965 committed:
#143 - Fixed Repair Button State Bug

Commit:
[#142] Styling Collapsible component for exercise status report

Commit:
[#142] Working on status report

jjm8759 committed:
Fixed eslint errors for lab7

notjoncruz committed:
Merge pull request #195 from all-rit/143-improving-the-autonomous-system-exercise

143 improving the autonomous system exercise

Commit:
Merge branch '143-improving-the-autonomous-system-exercise' into 142-simulation-exercise

# Conflicts:
#	client/package-lock.json
#	client/src/assets/stylesheets/components/AutoSysAI.scss
#	client/src/assets/stylesheets/components/File.scss
#	client/src/assets/stylesheets/components/Files.scss
#	client/src/assets/stylesheets/components/Simulation.scss
#	client/src/components/exercise/lab7/components/AutoSysAI.js
#	client/src/components/exercise/lab7/components/Collapsible.js
#	client/src/components/exercise/lab7/components/File.js
#	client/src/components/exercise/lab7/components/Files.js
#	client/src/components/exercise/lab7/components/Message.js
#	client/src/components/exercise/lab7/components/Simulation.js
#	client/src/components/exercise/lab7/components/mockData/fileMockData.js
#	client/src/components/exercise/lab7/pages/Simulation/AISimulation.js
#	client/src/components/exercise/lab7/pages/Simulation/SimulationSummary.js
#	client/src/constants/lab7/index.js
#	client/src/reducers/lab7/ExerciseReducer.js
#	client/src/services/lab7/ExerciseService.js

Commit:
[#142] Fixing errors and finalizing summary report

dm9965 committed:
[#144] Fixed State issue with Lab 7alteration quiz

Commit:
[#142] Added method documentation, finalized phase one and updated constants

dm9965 committed:
#144 Created prop to hide certificate on Alteration Quiz Result

dm9965 committed:
#200 updated quiz component to match updated config for Lab 7

Commit:
[#143] Decoupling code component

Commit:
[#143] Adding redux use in components and removing the passing of props

Commit:
[#143] More fixes with redux implementation

dm9965 committed:
#181 Updated Exercise End Pages for Lab 3, 4, 5

Commit:
[#143] Fixing styles on Lab 7 and allowing phase two to be taken multiple times

Commit:
[#143] Added functional user input parsing for utility equation to run simulation.

Commit:
[#142] Simulation styling

Commit:
[#143] Removed unused style file

Commit:
Merging #142, #143, #144

Commit:
colorPicker warning removed

jjm8759 commented:
15 seconds for the simulation seems a little bit long. During class time that could soak up a lot of time. Thoughts? @Mster115

Mster115 commented:
> 15 seconds for the simulation seems a little bit long. During class time that could soak up a lot of time. Thoughts? @Mster115

@jjm8759 I brought this up could be brought down to 10, it should be enough time for someone to read especially with any type of disability

jjm8759 commented:
> > 15 seconds for the simulation seems a little bit long. During class time that could soak up a lot of time. Thoughts? @Mster115
> 
> @jjm8759 I brought this up could be brought down to 10, it should be enough time for someone to read especially with any type of disability

Ok cool, I'll let Jon know.

Commit:
[#216] Decreasing read time to 10s from 15s

notjoncruz commented:
> > 15 seconds for the simulation seems a little bit long. During class time that could soak up a lot of time. Thoughts? @Mster115
> 
> @jjm8759 I brought this up could be brought down to 10, it should be enough time for someone to read especially with any type of disability

Fixed

Mster115 commented:
I noticed i couldnt get the repair working, it said my input was incorrect. Also the tooltip is inconsistent with the comment i believe tooltip said file.sensitivitylvl while comment is file.sensitivitylevel

notjoncruz commented:
> I noticed i couldnt get the repair working, it said my input was incorrect. Also the tooltip is inconsistent with the comment i believe tooltip said file.sensitivitylvl while comment is file.sensitivitylevel

Do you remember your input? Would like to test this out and look into it.

Mster115 commented:
> > I noticed i couldnt get the repair working, it said my input was incorrect. Also the tooltip is inconsistent with the comment i believe tooltip said file.sensitivitylvl while comment is file.sensitivitylevel
> 
> Do you remember your input? Would like to test this out and look into it.

I believe it was the exact input from the comment above minus the quotes

Commit:
Updating state at mount for each component

Commit:
Updating Alteration Quiz to display equations using latex

Commit:
Package changes

Commit:
Input title updated

notjoncruz commented:
> I noticed i couldnt get the repair working, it said my input was incorrect. Also the tooltip is inconsistent with the comment i believe tooltip said file.sensitivitylvl while comment is file.sensitivitylevel

I am unable to replicate the error. There was only one reference to `file.sensitivitylv` but that was the title of the input. Can you pull the changes and let me know if you still get this error?

Commit:
Fixed popup color timeout bug

dm9965 commented:
- In the "Unsupervised Learning" section of the reading, it says there's 2 types: "Clustering and Association", but then goes on to describe "Clustering and Classification"
- The faster file intrusion description time is much better and helps the flow of the lab
- The lab is slightly different across safari and chrome, the grey of the round results is darker in safari vs chrome.
- The lab does not mention intrusion description in the lab run after the repair has been made, not sure if that is intentional or not
- End of lab should have a "Return to Home" button, like with the rest of the labs. Likely a repair that can be made across all labs.

Mster115 commented:
> > I noticed i couldnt get the repair working, it said my input was incorrect. Also the tooltip is inconsistent with the comment i believe tooltip said file.sensitivitylvl while comment is file.sensitivitylevel
> 
> I am unable to replicate the error. There was only one reference to `file.sensitivitylv` but that was the title of the input. Can you pull the changes and let me know if you still get this error?

<img width="1117" alt="Screenshot 2023-02-19 at 3 38 33 PM" src="https://user-images.githubusercontent.com/69928357/219974006-7f8c520b-9f9d-478d-8d72-33136a56b956.png">

<img width="1098" alt="Screenshot 2023-02-19 at 3 38 21 PM" src="https://user-images.githubusercontent.com/69928357/219974017-46f46a59-bfd0-4d2a-beb9-7ff962493442.png">

Repair now works, issue is its not apparent when there is an issue because its always marked as wrong, note other labs it states this after user submits and its wrong. 

<img width="587" alt="Screenshot 2023-02-19 at 3 42 43 PM" src="https://user-images.githubusercontent.com/69928357/219974116-1dbfe380-cdc7-4642-af68-087a78e0d610.png">

user should be notified after submit the answers wrong (ie. box red and it states what it should be) let me know if you need more clarification @notcruz

Mster115 commented:
<img width="989" alt="Screenshot 2023-02-19 at 3 48 30 PM" src="https://user-images.githubusercontent.com/69928357/219974349-5f411582-1474-4da6-a8ba-9a640b3fdc6a.png">
 no third equation its blank? intentional? if it is i think it needs something

Mster115 commented:
> <img alt="Screenshot 2023-02-19 at 3 48 30 PM" width="989" src="https://user-images.githubusercontent.com/69928357/219974349-5f411582-1474-4da6-a8ba-9a640b3fdc6a.png">
> 
> no third equation its blank? intentional? if it is i think it needs something

to follow up with this ending seems abrupt maybe add a button saying finish exercise that returns you to the start of the exercise or right to reinforcement

Mster115 commented:
Also with the quiz we just updated the quiz component. You will need to make it work with quiz handler. with that though you may need to modify how it works because on complete it updates the database as a quiz however this isnt a quiz. this is a quiz type functionality but we still want the behavior saved in the lab7 exercise database

Mster115 commented on client/src/components/quiz/App.js:
```diff
@@ -11,13 +11,16 @@ import quizQuestionsLab2 from "./api/Lab2/quizQuestions";
 import quizQuestionsLab3 from "./api/Lab3/quizQuestions";
 import quizQuestionsLab4 from "./api/Lab4/quizQuestions";
 import quizQuestionsLab5 from "./api/Lab5/quizQuestions";
+import alterationQuizQuestions from "./api/Lab7/alterationQuizQuestions";
```
discussed above

notjoncruz replied:
Done.

Mster115 commented on client/src/services/lab7/ExerciseService.js:
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable react/prop-types */
+/* eslint-disable no-unused-vars */
+import API from "../API";
```
is this used? should be for user input

notjoncruz replied:
Done.

Mster115 commented on client/src/services/lab7/RepairService.js:
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable react/prop-types */
```
same here, repair is separate from exercise input

notjoncruz replied:
Done.

Mster115 reviewed (changes requested):
Looks good, some changes need to be addressed.

Commit:
[#216] Fixed question three not appearing on alteration quiz

notjoncruz commented:
> to follow up with this ending seems abrupt maybe add a button saying finish exercise that returns you to the start of the exercise or right to reinforcement

I was testing latex and forgot to remove the change. Fixed.

notjoncruz commented:
> > <img alt="Screenshot 2023-02-19 at 3 48 30 PM" width="989" src="https://user-images.githubusercontent.com/69928357/219974349-5f411582-1474-4da6-a8ba-9a640b3fdc6a.png">
> > no third equation its blank? intentional? if it is i think it needs something
> 
> to follow up with this ending seems abrupt maybe add a button saying finish exercise that returns you to the start of the exercise or right to reinforcement

The buttons at the bottom appear once the alteration quiz is complete. Are these the options you are referring to?

![image](https://user-images.githubusercontent.com/70080228/219980088-894f1412-b02b-48c4-9a8d-c3ad7f22cab1.png)

Mster115 commented:
> > > <img alt="Screenshot 2023-02-19 at 3 48 30 PM" width="989" src="https://user-images.githubusercontent.com/69928357/219974349-5f411582-1474-4da6-a8ba-9a640b3fdc6a.png">
> > > no third equation its blank? intentional? if it is i think it needs something
> > 
> > 
> > to follow up with this ending seems abrupt maybe add a button saying finish exercise that returns you to the start of the exercise or right to reinforcement
> 
> The buttons at the bottom appear once the alteration quiz is complete. Are these the options you are referring to?
> 
> ![image](https://user-images.githubusercontent.com/70080228/219980088-894f1412-b02b-48c4-9a8d-c3ad7f22cab1.png)

<img width="956" alt="Screenshot 2023-02-19 at 6 00 37 PM" src="https://user-images.githubusercontent.com/69928357/219980625-4a12e6ff-5fc6-42d8-887e-e58f58d8d9f6.png">

I mean adding a button below the end of the quiz just to make it prominent its over similar to this

Commit:
[#216] Updated code repair to inform user of error

notjoncruz commented:
> > > I noticed i couldnt get the repair working, it said my input was incorrect. Also the tooltip is inconsistent with the comment i believe tooltip said file.sensitivitylvl while comment is file.sensitivitylevel
> > 
> > 
> > I am unable to replicate the error. There was only one reference to `file.sensitivitylv` but that was the title of the input. Can you pull the changes and let me know if you still get this error?
> 
> <img alt="Screenshot 2023-02-19 at 3 38 33 PM" width="1117" src="https://user-images.githubusercontent.com/69928357/219974006-7f8c520b-9f9d-478d-8d72-33136a56b956.png">
> 
> <img alt="Screenshot 2023-02-19 at 3 38 21 PM" width="1098" src="https://user-images.githubusercontent.com/69928357/219974017-46f46a59-bfd0-4d2a-beb9-7ff962493442.png">
> 
> Repair now works, issue is its not apparent when there is an issue because its always marked as wrong, note other labs it states this after user submits and its wrong.
> 
> <img alt="Screenshot 2023-02-19 at 3 42 43 PM" width="587" src="https://user-images.githubusercontent.com/69928357/219974116-1dbfe380-cdc7-4642-af68-087a78e0d610.png">
> 
> user should be notified after submit the answers wrong (ie. box red and it states what it should be) let me know if you need more clarification @notcruz

Fixed.

AndreasLc1103 commented on client/src/components/exercise/lab7/Main.js:
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable react/prop-types */
+/* eslint-disable require-jsdoc */
```
Fill in doc strings on this

notjoncruz replied:
Resolved. No function other than render.

AndreasLc1103 commented on client/src/components/exercise/lab7/Main.js:
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable react/prop-types */
```
required proptypes would also make this a little bit easier to track 
https://reactjs.org/docs/typechecking-with-proptypes.html

notjoncruz replied:
Resolved.

AndreasLc1103 commented on client/src/components/exercise/lab7/components/Code.js:
```diff
@@ -0,0 +1,347 @@
+} from "../../../../constants/lab7";
+import { evaluate } from "mathjs";
+
+class Code extends Component {
```
if this is gonna be a component I would make it a functional one

notjoncruz replied:
Talked in person. Better to keep as class.

AndreasLc1103 commented on client/src/components/exercise/lab7/components/Code.js:
```diff
@@ -0,0 +1,347 @@
+            <div className="code_editor__file code_editor__file--active">
+              AutoSysAI.js
+            </div>
+            {/* psuedocode */}
```
remove commented out psuedocode

notjoncruz replied:
Fixed.

heathermoses commented:
![Screenshot 2023-02-20 at 9 46 53 AM](https://user-images.githubusercontent.com/54957497/220137682-239a5de7-2f0e-4b93-b93e-603d8f972702.png)
Not a make or break change but Jon, Janae, and Dom's names should be added as authors of this lab :)

Commit:
[#216] Proper use of react-redux to components

Commit:
[#216] Handled all es-lint ignores

notjoncruz commented:
> ![Screenshot 2023-02-20 at 9 46 53 AM](https://user-images.githubusercontent.com/54957497/220137682-239a5de7-2f0e-4b93-b93e-603d8f972702.png) Not a make or break change but Jon, Janae, and Dom's names should be added as authors of this lab :)

Added.

2nPlusOne reviewed (commented):
# Lab 7 Review

## More Info

Nice job with this. It is very clear and concise.

- Would this sound better as either "through a cybersecurity lens." or "with a focus on cybersecurity." ?
<img width="641" alt="image" src="https://user-images.githubusercontent.com/39078738/221383848-3a660ef5-7a92-4d13-9c1e-7ef563dcf6f1.png">


## About

Looks good! I just noticed a missing comma (below).

<img width="875" alt="image" src="https://user-images.githubusercontent.com/39078738/221383937-e1c17982-2ffd-4fb6-a717-9c16783b105a.png">


## Reading

Excellent job with the reading section, I found it incredibly informative. I only have a couple of small comments.
- Would this sound better without the second "to you" ? ie, "Whether cybersecurity seems interesting to you, or irrelevant, anyone who uses the internet ..."
<img width="886" alt="image" src="https://user-images.githubusercontent.com/39078738/221384052-416095d6-fbf0-451a-a1c1-787b4f5188f1.png">

- Great statistic! 
<img width="870" alt="image" src="https://user-images.githubusercontent.com/39078738/221384142-ce9b7b1c-239e-4c71-9366-0498d88b67ce.png">

With the rise of AI chatbots like ChatGPT, scammers can quickly and cheaply produce more convincing phishing emails and webpages. It might be worth including something about this in the reading section.

- This would be binary classification. There are other forms of classification, such as multi-class classification. This might be rephrased as: "The algorithm categorizes the output into distinct classes: yes or no, true or false, or a letter of the alphabet (in a letter recognition model)."
<img width="633" alt="image" src="https://user-images.githubusercontent.com/39078738/221384405-28b5353d-9481-444c-9909-4d565b8ccb46.png">

## Exercise

### Instructions

The instructions are rather dense. I think they would benefit from some white space between nested bullet sections.

- Is this meant to say 'sensitivity level'?
<img width="805" alt="image" src="https://user-images.githubusercontent.com/39078738/221385092-1a2af848-df92-46b4-818b-ea3aea0e442a.png">

### Exercise

- These were all home address, resulting in the message staying the same each time.
<img width="1067" alt="image" src="https://user-images.githubusercontent.com/39078738/221384840-3f5aed03-0704-4c13-8292-fbea8aa8286d.png">

- Clicking update doesn't appear to do anything. It was only when I scrolled up that I saw the message that the fixed was applied. This could be fixed by having the page scroll up when the fix is successful.
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/39078738/221384949-14f666b3-bed6-4ed5-a745-f11889729357.png">

I don't see the utility of going through the repair section multiple times since the inputs only ask for a specific value. Is there something I'm missing?

After applying the fix, the AI was still marked as having made many mistakes. Overall I had a hard time grokking how the activity worked, specifically around the connection between a file's sensitivity level and whether it should be open or locked for a particular threat level.

## Reinforcement

The last two reinforcement videos are around 18 minutes long. Because of this, the likelihood that a user will follow through and watch them is low. Here are a couple alternatives that might be considered:

- [Artificial Intelligence for smarter Cybersecurity](https://www.youtube.com/watch?v=rH9-m7AhJhk)
- [Applying Deep Learning To Detect Any Cyber-Threat](https://www.youtube.com/watch?v=_qdvTAgKiv0)

## Quiz

- These two sentences are missing a subject ("models"). 
<img width="749" alt="image" src="https://user-images.githubusercontent.com/39078738/221385778-8f5ad96b-7b3c-4b87-abdb-f3af2919f0ce.png">
- "Machine learning models are expensive to develop and to use"
- "Machine learning models need to perform well to replace humans in the workplace"

Mster115 committed:
Merge branch 'dev' into development/Lab-7

Mster115 committed:
prettier fix

Mster115 committed:
stubbed func for non final quiz data submission

Mster115 committed:
prettier fix

Commit:
[#216] DB Implementation

Commit:
[#216] DB Implementation

Commit:
[#216] Alteration Quiz DB implementation

AndreasLc1103 commented on client/src/components/exercise/lab7/Main.js:
```diff
@@ -0,0 +1,42 @@
+import AlterationQuiz from "./pages/AlterationActivity/AlterationQuiz";
+import ExerciseEnd from "./pages/ExerciseEnd";
+
+class Main extends Component {
```
I would make this a functional component if you can. should be a quick fix and shouldn't interfere with functionality

notjoncruz replied:
Updated.

Mster115 commented:
Small change before merge, can the repair close after update button is submitted successfully to make next button more prominent. Also follows format of other labs

Mster115 commented:
<img width="1218" alt="Screenshot 2023-02-28 at 8 38 57 AM" src="https://user-images.githubusercontent.com/69928357/221870419-3f335e86-31ff-4b0e-a7cd-d02ca7983638.png">
Also could you add a button at the bottom that allows transition to an exercise end wrapping up what they learned and turning gamestate back to idle so they can navigate the site.

Mster115 commented:
> <img alt="Screenshot 2023-02-28 at 8 38 57 AM" width="1218" src="https://user-images.githubusercontent.com/69928357/221870419-3f335e86-31ff-4b0e-a7cd-d02ca7983638.png">
> 
> Also could you add a button at the bottom that allows transition to an exercise end wrapping up what they learned and turning gamestate back to idle so they can navigate the site.

At the end it should also have a button that returns the exercise to the exercise start page, or to the reinforcement page whatever makes the most sense.

Mster115 reviewed (changes requested):
Final remarks above

AndreasLc1103 commented on client/src/services/lab7/RepairService.js:
```diff
@@ -0,0 +1,25 @@
+const RepairService = {
+  submitRepair: (activity, repair) => {
+    return API.postWithBody(
+      process.env.REACT_APP_SERVER_URL + endpoints.SUBMIT_REPAIR,
```
```suggestion
      `${process.env.REACT_APP_SERVER_URL}${endpoints.SUBMIT_REPAIR}`,
```

AndreasLc1103 replied:
This applies to any of the services that are included within the functionality of this lab

notjoncruz replied:
Updated

AndreasLc1103 reviewed (commented):
Some minor fixes. also, could you include it in the git ignore to ignore changes to the eslintrcs

Commit:
[#216] Ryan's review implementation

notjoncruz commented:
> # Lab 7 Review
> ## More Info
> Nice job with this. It is very clear and concise.
> 
> * Would this sound better as either "through a cybersecurity lens." or "with a focus on cybersecurity." ?
> 
> <img alt="image" width="641" src="https://user-images.githubusercontent.com/39078738/221383848-3a660ef5-7a92-4d13-9c1e-7ef563dcf6f1.png">
> 
> ## About
> Looks good! I just noticed a missing comma (below).
> 
> <img alt="image" width="875" src="https://user-images.githubusercontent.com/39078738/221383937-e1c17982-2ffd-4fb6-a717-9c16783b105a.png">
> 
> ## Reading
> Excellent job with the reading section, I found it incredibly informative. I only have a couple of small comments.
> 
> * Would this sound better without the second "to you" ? ie, "Whether cybersecurity seems interesting to you, or irrelevant, anyone who uses the internet ..."
> 
> <img alt="image" width="886" src="https://user-images.githubusercontent.com/39078738/221384052-416095d6-fbf0-451a-a1c1-787b4f5188f1.png">
> 
> * Great statistic!
> 
> <img alt="image" width="870" src="https://user-images.githubusercontent.com/39078738/221384142-ce9b7b1c-239e-4c71-9366-0498d88b67ce.png">
> 
> With the rise of AI chatbots like ChatGPT, scammers can quickly and cheaply produce more convincing phishing emails and webpages. It might be worth including something about this in the reading section.
> 
> * This would be binary classification. There are other forms of classification, such as multi-class classification. This might be rephrased as: "The algorithm categorizes the output into distinct classes: yes or no, true or false, or a letter of the alphabet (in a letter recognition model)."
> 
> <img alt="image" width="633" src="https://user-images.githubusercontent.com/39078738/221384405-28b5353d-9481-444c-9909-4d565b8ccb46.png">
> 
> ## Exercise
> ### Instructions
> The instructions are rather dense. I think they would benefit from some white space between nested bullet sections.
> 
> * Is this meant to say 'sensitivity level'?
> 
> <img alt="image" width="805" src="https://user-images.githubusercontent.com/39078738/221385092-1a2af848-df92-46b4-818b-ea3aea0e442a.png">
> 
> ### Exercise
> * These were all home address, resulting in the message staying the same each time.
> 
> <img alt="image" width="1067" src="https://user-images.githubusercontent.com/39078738/221384840-3f5aed03-0704-4c13-8292-fbea8aa8286d.png">
> 
> * Clicking update doesn't appear to do anything. It was only when I scrolled up that I saw the message that the fixed was applied. This could be fixed by having the page scroll up when the fix is successful.
> 
> <img alt="image" width="1280" src="https://user-images.githubusercontent.com/39078738/221384949-14f666b3-bed6-4ed5-a745-f11889729357.png">
> 
> I don't see the utility of going through the repair section multiple times since the inputs only ask for a specific value. Is there something I'm missing?
> 
> After applying the fix, the AI was still marked as having made many mistakes. Overall I had a hard time grokking how the activity worked, specifically around the connection between a file's sensitivity level and whether it should be open or locked for a particular threat level.
> 
> ## Reinforcement
> The last two reinforcement videos are around 18 minutes long. Because of this, the likelihood that a user will follow through and watch them is low. Here are a couple alternatives that might be considered:
> 
> * [Artificial Intelligence for smarter Cybersecurity](https://www.youtube.com/watch?v=rH9-m7AhJhk)
> * [Applying Deep Learning To Detect Any Cyber-Threat](https://www.youtube.com/watch?v=_qdvTAgKiv0)
> 
> ## Quiz
> * These two sentences are missing a subject ("models").
> 
> <img alt="image" width="749" src="https://user-images.githubusercontent.com/39078738/221385778-8f5ad96b-7b3c-4b87-abdb-f3af2919f0ce.png">
> 
> - "Machine learning models are expensive to develop and to use" - "Machine learning models need to perform well to replace humans in the workplace"

Thanks for the review! Updated the lab accordingly.

notjoncruz commented:
> Small change before merge, can the repair close after update button is submitted successfully to make next button more prominent. Also follows format of other labs

Updated functionality

notjoncruz commented:
> > <img alt="Screenshot 2023-02-28 at 8 38 57 AM" width="1218" src="https://user-images.githubusercontent.com/69928357/221870419-3f335e86-31ff-4b0e-a7cd-d02ca7983638.png">
> > Also could you add a button at the bottom that allows transition to an exercise end wrapping up what they learned and turning gamestate back to idle so they can navigate the site.
> 
> At the end it should also have a button that returns the exercise to the exercise start page, or to the reinforcement page whatever makes the most sense.

Updated and Fixed

Commit:
[#216] Andreas' review implementation

Mster115 commented:
> > > <img alt="Screenshot 2023-02-28 at 8 38 57 AM" width="1218" src="https://user-images.githubusercontent.com/69928357/221870419-3f335e86-31ff-4b0e-a7cd-d02ca7983638.png">
> > > Also could you add a button at the bottom that allows transition to an exercise end wrapping up what they learned and turning gamestate back to idle so they can navigate the site.
> > 
> > 
> > At the end it should also have a button that returns the exercise to the exercise start page, or to the reinforcement page whatever makes the most sense.
> 
> Updated and Fixed

@notcruz is there a button that goes to an exercise end page? or a return to beginning of exercise?

notjoncruz commented:
> > > > <img alt="Screenshot 2023-02-28 at 8 38 57 AM" width="1218" src="https://user-images.githubusercontent.com/69928357/221870419-3f335e86-31ff-4b0e-a7cd-d02ca7983638.png">
> > > > Also could you add a button at the bottom that allows transition to an exercise end wrapping up what they learned and turning gamestate back to idle so they can navigate the site.
> > > 
> > > 
> > > At the end it should also have a button that returns the exercise to the exercise start page, or to the reinforcement page whatever makes the most sense.
> > 
> > 
> > Updated and Fixed
> 
> @notcruz is there a button that goes to an exercise end page? or a return to beginning of exercise?

Yeah, the state is updated to `EXERCISE_IDLE`. Are you running on the latest changes? I'll have Dom check it out as well.

![image](https://user-images.githubusercontent.com/70080228/222838210-ae120ad8-1c37-4096-b60b-6a10d4a98aea.png)

Mster115 commented:
> > > > > <img alt="Screenshot 2023-02-28 at 8 38 57 AM" width="1218" src="https://user-images.githubusercontent.com/69928357/221870419-3f335e86-31ff-4b0e-a7cd-d02ca7983638.png">
> > > > > Also could you add a button at the bottom that allows transition to an exercise end wrapping up what they learned and turning gamestate back to idle so they can navigate the site.
> > > > 
> > > > 
> > > > At the end it should also have a button that returns the exercise to the exercise start page, or to the reinforcement page whatever makes the most sense.
> > > 
> > > 
> > > Updated and Fixed
> > 
> > 
> > @notcruz is there a button that goes to an exercise end page? or a return to beginning of exercise?
> 
> Yeah, the state is updated to `EXERCISE_IDLE`. Are you running on the latest changes? I'll have Dom check it out as well.
> 
> ![image](https://user-images.githubusercontent.com/70080228/222838210-ae120ad8-1c37-4096-b60b-6a10d4a98aea.png)

@notcruz addressed offline waiting for change

stthazin commented on client/src/constants/lab7/index.js:
```diff
@@ -0,0 +1,60 @@
+  "Home Address":
+    "Someone’s home address was stolen during the security breach. As a result of the breach, an identity thief was able to change the individual’s mailing address and reroute their mail to another address. This autonomous system decision-mistake caused a case of address fraud.",
+  "Mother Maiden Name":
+    "Someone’s mother's maiden was stolen during the security breach. As a result of the breach, an identity thief was able to answer the individual's security questions for all their accounts. This autonomous system decision-mistake caused a case of hacking and identity fraud.",
```
I think you meant to say "maiden _name_"

stthazin commented on client/src/constants/lab7/index.js:
```diff
@@ -0,0 +1,60 @@
+  "Mother Maiden Name":
+    "Someone’s mother's maiden was stolen during the security breach. As a result of the breach, an identity thief was able to answer the individual's security questions for all their accounts. This autonomous system decision-mistake caused a case of hacking and identity fraud.",
+  "Full Name":
+    "Someone’s full was stolen during the security breach. As a result of the breach, an identity thief was able to sign up and create fake accounts under the individual's name. This autonomous system decision-mistake caused a case of identity fraud.",
```
same thing here - "full _name_"

stthazin commented on client/src/constants/lab7/index.js:
```diff
@@ -0,0 +1,60 @@
+export const THREAT_LEVEL_TEXT = { 1: "Low", 2: "Medium", 3: "High" };
+
+export const MESSAGES = {
+  SSN: "Someone’s social security number was stolen during the security breach. As a result of the breach, an identity thief was able to steal an individual's identity. This autonomous system decision-mistake caused a case of indentity theft.",
```
typo in "identity" at the end

stthazin reviewed (changes requested):
I just wanted to point out some typos - very brief review

stthazin commented on client/src/components/exercise/lab7/components/File.js:
```diff
@@ -0,0 +1,59 @@
+          </div>
+        )}
+        <div>
+          <p className={"tw-leading-tight tw-font-bold"}>{data.fileName}</p>
```
this is just my opinion but I don't think filename is too important in this context and especially when there's a lot of text, maybe this doesn't have to be bold?

notjoncruz replied:
Good point, removed the bold for the filename.

stthazin commented on client/src/components/exercise/lab7/pages/ExerciseStart.js:
```diff
@@ -0,0 +1,153 @@
+              </li>
+              <li>
+                At the end, a summary report containing all of the information
+                displayed throughout the simulation will be displayed
```
full stop at the end?

notjoncruz replied:
I'm a bit confused here, what do you mean?

Mster115 replied:
"." a period at the end I believe
@notcruz

stthazin commented on client/src/components/exercise/lab7/pages/ImproveAICode/AICodeRepair.js:
```diff
@@ -0,0 +1,153 @@
+              <p className={"playthrough__sentence"}>
+                Your goal should be to implement a utility equation into the
+                autonomous system. The utility equation is calculated by
+                weighting the reward of a decision and the cost of making said
```
"_weighing_" instead of "weighting"

stthazin commented on client/src/components/exercise/lab7/pages/Simulation/BadAIExplanation.js:
```diff
@@ -0,0 +1,63 @@
+        </p>
+        <p className="playthrough__sentence">
+          This is due to the factors of the files that the AI is using to
+          determine in a file&lsquo;s access should be restricted or not.
```
I think this is "determine _if_" instead of "in"?

stthazin commented on client/src/constants/lab7/index.js:
```diff
@@ -0,0 +1,60 @@
+
+export const POPUP_MESSAGES = {
+  INVALID_EXPRESSION: "You must pass a valid expression.",
+  ZERO_DIVISION: "Zero division is not allowed.",
```
"Division by zero" is the correct term I think

stthazin commented on client/src/constants/lab7/index.js:
```diff
@@ -0,0 +1,60 @@
+export const ROUND_LIMIT = 10;
+
+export const POPUP_MESSAGES = {
+  INVALID_EXPRESSION: "You must pass a valid expression.",
```
"pass _in_"

stthazin reviewed (changes requested):
Just some more grammatical things

Commit:
Updates based on demo

notjoncruz commented:
Looking into a bug that causes the modal div to still be on the Elements Tree even though the div is disabled.

Commit:
More updates

notjoncruz commented:
> Looking into a bug that causes the modal div to still be on the Elements Tree even though the div is disabled.

Found this [issue](https://github.com/reactstrap/reactstrap/issues/1323) online. I believe the bug came from the `setTimeout` invocations, which updated the states at the wrong times. I updated/fixed the timing, which has fixed the problem.

Mster115 committed:
Merge branch 'dev' into development/Lab-7

Mster115 committed:
fixed change effected by previous defect

Mster115 committed:
prettier fixes

Mster115 commented:
NOTES: For now everything seems fine, possible typos addressed in hotfix, and addressing repair page instructions

Mster115 committed:
fixed ending of exercise

Mster115 committed:
prettier fixes
